service/raft/raft_group0.cc

```diff
@@ -560,6 +560,51 @@ future<> raft_group0::setup_group0(db::system_keyspace& sys_ks, const std::unord
     co_await _client.set_group0_upgrade_state(group0_upgrade_state::use_post_raft_procedures);
 }
 
+future<> raft_group0::persist_initial_raft_address_map() {
+    static constexpr auto max_concurrency = 10;
+    std::vector<gms::inet_address> endpoints;
+    const auto& states = _gossiper.get_endpoint_states();
+    endpoints.reserve(states.size());
+    std::transform(states.begin(), states.end(), std::back_inserter(endpoints), [](const auto& it) { return it.first; });
+    //
+    // Gossiper only persists node state when it switches to
+    // NORMAL. When persisting a node state, Raft ID is only
+    // persisted if the cluster supports raft.  Some nodes switch
+    // to NORMAL before the whole cluster enables raft. For these
+    // nodes we may never persist their Raft ID unless, here, as
+    // soon as we discover that the entire cluster supports Raft,
+    // we go over these nodes and explicitly persist their RAFT
+    // IDs.
+
+    co_await max_concurrent_for_each(endpoints, max_concurrency, [this] (const gms::inet_address& ip_addr) -> future<> {
+        auto* state = _gossiper.get_endpoint_state_for_endpoint_ptr(ip_addr);
```

Comment:
Maybe we could do all of this processing before `co_await _sys_ks.update_peer_info` outside the `max_concurrent_for_each` - inside the `transform`? It's cheap, and the number of nodes is for all practical purposes < 100 (or < 1000, or whatever) so we can do it without yielding. The `max_concurrent_for_each` would go over a vector of `ip_addr, server_id` pairs.

But I guess it's mostly style choice.

Reply:
The whole point is that this list may change. So why ask for additional trouble:
- a higher chance of a reactor stall
- a higher chance to persist an obsolete entry